File: javascript/diamond/diamond.js
//
// This is only a SKELETON file for the 'Diamond' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const rows = (letter) => {
  const uppercases = new Array(26).fill(1).map((_, i) => String.fromCharCode(65 + i))

  if (!uppercases.includes(letter)) {
    throw new Error("letter is not a valid ascii uppercase")
  }

  const diamond = []
  const length = uppercases.indexOf(letter) * 2 + 1

  for (let i = 0; i < Math.ceil(length / 2); i++) {
    const letter_positions = i > 0 ? 
      [Math.floor(length / 2) - i, Math.floor(length / 2) + i] :
      [Math.floor(length / 2)]
    const row = []

    for (let j = 0; j < length; j++) {
      const point = letter_positions.includes(j) ? uppercases[i] : " "
      row.push(point)
    }

    diamond.push(row.join(''))
  }

  for (let i = Math.floor(length / 2); i > 0; i--) {
    const letter_positions = i > 1 ?
      [Math.ceil(length / 2) - i, Math.floor(length / 2) + i-1] :
      [Math.floor(length / 2)]
    const row = []

    for (let j = 0; j < length; j++) {
      const point = letter_positions.includes(j) ? uppercases[i-1] : " "
      row.push(point)
    }

    diamond.push(row.join(''))
  }

  return diamond
}

File: python/diamond/diamond.py
import string

def rows(letter: string) -> list:
    uppercases = list(string.ascii_uppercase)

    if letter not in uppercases:
        raise ValueError("letter is not a valid ascii uppercase")

    diamond = []
    length = uppercases.index(letter) * 2 + 1

    for line in range(length // 2 + 1):
        letter_pos = [length // 2 - line, length // 2 + line] if line > 0 else [length // 2]

        row = []
        for point_index in range(length):
            point = " " if point_index not in letter_pos else uppercases[line]
            row.append(point)

        diamond.append("".join(row))

    for line in range(length // 2, 0, -1):
        letter_pos = [length // 2 - line+1, length // 2 + line-1] if line > 1 else [length // 2]
        
        row = []
        for point_index in range(length):
            point = " " if point_index not in letter_pos else uppercases[line-1]
            row.append(point)

        diamond.append("".join(row))

    return diamond